### PR TITLE
#9301 Support to configured formatters and quick filters in feature grid

### LINKED ISSUE #9301: Custom renderers and quick-filters injection in FeatureEditor plugin by other plugins
## Acceptance Criteria

As a developer, I want to develop plugins that add some custom renderers to be used in all the cells of a given column of a given feature of the feature grid, and some custom quick filter tools to be used in the hader of a given column for a given feature. Actually this is supported only for editors, and they have to be registered using a proper Javascript utility. With this improvement also the editor could be injectable with the same logic. 
- I want that this renderer is available for use in general (e.g. for general purpose renderers and filters)
- I want to be able to configure also the plugin with rules similar to the custom editor, to use these renderers/filters using regular expressions to recognize feature types, URL and column names.
- I want to be able configure in layer fields to use these renderer and filter, when available, in the feature grid ( at least by editing configuration, without UI )

### Specific implementations

With this we want to provide support for custom project to some features like: 
- Show JSON attributes (represented as text by the describeFeatureType) in a clickable popup, that shows all the fields (for the particular project we will have some information about the structure of the JSON).
- Allow to filter these JSON attributes using CQL filters (e.g. jsonArrayContains(...), jsonPointer(...) in the column.

**What kind of improvement you want to add?** (check one with "x", remove the others)
- [x] Minor changes to existing features
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Other... Please describe:

## Other useful information

This is a feature that interest only custom projects. There should not be any UI change to explicitly test.

### PR DESCRIPTION
**Pull request by offtherailz**
## Description

Pending tasks: 
- [x] clean up
- [x] support for formatters
- [x] support for filterRenderers (quick filters)

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/geosolutions-it/MapStore2/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x", remove the others)
 - [ ] Bugfix
 - [x] Feature
 - [ ] Code style update (formatting, local variables)
 - [ ] Refactoring (no functional changes, no api changes)
 - [ ] Build related changes
 - [ ] CI related changes
 - [ ] Other... Please describe:

## Issue

**What is the current behavior?**

#9301 

**What is the new behavior?**
Fix #9301 

## Breaking change
**Does this PR introduce a breaking change?** (check one with "x", remove the other)
 - [ ] Yes, and I documented them in migration notes
 - [x] No

## Other useful information

No UI tests are needed for this feature. Only for custom projects support (and future improvements)

### CODE REVIEW

File: web/client/components/data/featuregrid/formatters/index.js [hunk 1]
```diff
@@ -13,41 +13,86 @@ import moment from "moment";
...
-        return ({value} = {}) => {
-            return !isNil(value)
-                ? moment.utc(value).isValid() // geoserver sometimes returns UTC for time.
-                    ? moment.utc(value).format(format)
-                    : moment(value).format(format)
-                : null;
+/**
+ * Registers a formatter component with the given name.
+ * @param {string} name the name of the formatter
+ * @param {component} formatter the formatter component
```
Comment (MV88) [inline web/client/components/data/featuregrid/formatters/index.js:37]:
```suggestion
 * @param {React.Component} formatter the formatter component
```

File: web/client/plugins/featuregrid/FeatureEditor.jsx [hunk 2]
```diff
@@ -187,7 +187,7 @@ const FeatureDock = (props = {
     };
     const items = props?.items ?? [];
     const toolbarItems = items.filter(({target}) => target === 'toolbar');
-
+    const filterRenderers = useMemo(() => getFilterRenderers(props.describe, props.fields), [props.describe, props.fields]);
```
Comment (MV88) [inline web/client/plugins/featuregrid/FeatureEditor.jsx:190]: good use of useMemo

File: web/client/plugins/widgets/getWidgetFilterRenderers.js [hunk 3]
```diff
@@ -35,10 +35,10 @@ const mergeQuickFiltersStream = compose(
 
 );
 
-const getFilterComponent = ({options, localType, attributeName, quickFilterStream$}) => compose(
+const getFilterComponent = ({field = {}, options, localType, attributeName, quickFilterStream$}) => compose(
     withProps({quickFilterStream$, attributeName, options}),
     mergeQuickFiltersStream
-)(getFilterRenderer(localType, {name: attributeName }));
+)(getFilterRenderer({name: field.filterRenderer?.name, type: localType, options: field.filterRenderer?.options}));
```
Comment (MV88) [inline web/client/plugins/widgets/getWidgetFilterRenderers.js:41]: you don't need anymore the attributeName  OR localType?

can they be added in OR as second condition?

**Review by MV88 — CHANGES_REQUESTED** [review]
overall looks good. just a few comments here and there.

this has to be tested in the custom project

File: web/client/plugins/widgets/getWidgetFilterRenderers.js (hunk 3 above)
Comment (offtherailz) [reply web/client/plugins/widgets/getWidgetFilterRenderers.js:41]: The localType is the default so it is needed. The name was never used in practice, So I removed to clean up the API

**Review by MV88 — APPROVED** [review]

**Pull request merged**